starlite/kwargs.py

```diff
@@ -409,39 +418,94 @@ def _collect_params(params: Mapping[str, Any], expected: Set[ParameterDefinition
             raise ValidationException(f"Missing required parameter(s) {', '.join(missing_params)} for url {url}")
         return {p.field_name: params.get(p.field_alias, p.default_value) for p in expected}
 
+    async def resolve_dependencies(self, connection: Union["WebSocket", "Request"], kwargs: Dict[str, Any]) -> None:
+        """Resolve all dependencies into the kwargs, recursively.
+
+        Args:
+            connection: An instance of [Request][starlite.connection.Request] or [WebSocket][starlite.connection.WebSocket].
+            kwargs: Kwargs to pass to dependencies
+        """
+        for batch in self.dependency_batches:
+            await self._resolve_dependencies_concurrently(batch, connection, kwargs)
+
+    async def _resolve_dependencies_concurrently(
+        self,
+        dependencies: Collection["Dependency"],
+        connection: Union["WebSocket", "Request"],
+        kwargs: Dict[str, Any],
+    ) -> None:
+        """Resolve the dependencies concurrently. The passed in kwargs will be modified with the results.
+
+        Args:
+            dependencies: An iterable for instances of [Dependency][starlite.kwargs.Dependency].
+            connection: An instance of [Request][starlite.connection.Request] or [WebSocket][starlite.connection.WebSocket].
+            kwargs: Kwargs to pass to dependencies
+        """
+        if len(dependencies) == 0:
+            pass
+        elif len(dependencies) == 1:
+            await self._resolve_dependency(next(iter(dependencies)), connection, kwargs)
+        else:
+            async with create_task_group() as task_group:
+                for dependency in dependencies:
+                    task_group.start_soon(self._resolve_dependency, dependency, connection, kwargs)
+
+    @staticmethod
     async def _resolve_dependency(
-        self, dependency: "Dependency", connection: Union["WebSocket", "Request"], **kwargs: Any
-    ) -> Any:
-        """Recursively resolve the dependency graph."""
+        dependency: "Dependency", connection: Union["WebSocket", "Request"], kwargs: Dict[str, Any]
+    ) -> None:
+        """Resolve a given instance of [Dependency][starlite.kwargs.Dependency]. All required sub dependencies must
+        already be resolved into the kwargs.
+
+        Args:
+            dependency: An instance of [Dependency][starlite.kwargs.Dependency]
+            connection: An instance of [Request][starlite.connection.Request] or [WebSocket][starlite.connection.WebSocket].
+            kwargs: Any kwargs to pass recursively.
+        """
         signature_model = get_signature_model(dependency.provide)
-        await resolve_dependencies_concurrently(self, dependency.dependencies, connection, kwargs)
         dependency_kwargs = signature_model.parse_values_from_connection_kwargs(connection=connection, **kwargs)
-        return await dependency.provide(**dependency_kwargs)
+        kwargs[dependency.key] = await dependency.provide(**dependency_kwargs)
```

Comment:
dependency.provide is not always async. Its important to note this - this is configured by the user.

Reply:
Not sure I understand, Provide.__call__ is async?